src/Analyzers/Core/Analyzers/RemoveUnnecessarySuppressions/SuppressMessageAttributeState.cs

```diff
@@ -0,0 +1,152 @@
+        {
+            _compilation = compilation;
+            _suppressMessageAttributeType = suppressMessageAttributeType;
+            _lazySuppressMessageAttributesBySyntax = new Lazy<ImmutableDictionary<SyntaxNode, AttributeData>>(CreateAttributesBySyntaxMap);
+        }
+
+        private static ImmutableDictionary<string, TargetScope> CreateTargetScopesMap()
+        {
+            var builder = ImmutableDictionary.CreateBuilder<string, TargetScope>(StringComparer.OrdinalIgnoreCase);
+
+#pragma warning disable CS8605 // Unboxing a possibly null value.
+            foreach (TargetScope targetScope in Enum.GetValues(typeof(TargetScope)))
+#pragma warning restore CS8605 // Unboxing a possibly null value.
+            {
+                if (targetScope == TargetScope.None)
+                {
+                    continue;
+                }
+
+                builder.Add(targetScope.ToString(), targetScope);
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private ImmutableDictionary<SyntaxNode, AttributeData> CreateAttributesBySyntaxMap()
+        {
+            var builder = ImmutableDictionary.CreateBuilder<SyntaxNode, AttributeData>();
+            AddAttributes(_compilation.Assembly, _suppressMessageAttributeType, builder);
+            foreach (var module in _compilation.Assembly.Modules)
+            {
+                AddAttributes(module, _suppressMessageAttributeType, builder);
+            }
+
+            return builder.ToImmutable();
+
+            // Local functions.
+            static void AddAttributes(ISymbol symbol, INamedTypeSymbol suppressMessageAttributeType, ImmutableDictionary<SyntaxNode, AttributeData>.Builder builder)
+            {
+                foreach (var attribute in symbol.GetAttributes())
+                {
+                    if (suppressMessageAttributeType.Equals(attribute.AttributeClass) &&
+                        attribute.ApplicationSyntaxReference?.GetSyntax() is SyntaxNode node)
+                    {
+                        builder.Add(node, attribute);
+                    }
+                }
+            }
+        }
+
+        public bool IsGlobalSuppressMessageAttribute(SyntaxNode attributeSyntax, [NotNullWhen(returnValue: true)] out AttributeData? attribute)
+            => _lazySuppressMessageAttributesBySyntax.Value.TryGetValue(attributeSyntax, out attribute);
+
+        public bool HasInvalidScope(AttributeData attribute, out TargetScope targetScope)
+        {
+            if (!TryGetNamedArgument(attribute, SuppressMessageScope, out var scopeString) ||
+                string.IsNullOrEmpty(scopeString))
+            {
+                // Missing/Null/Empty scope values are treated equivalent to a compilation wide suppression.
+                targetScope = TargetScope.Module;
+            }
+            else if (!s_targetScopesMap.TryGetValue(scopeString!, out targetScope))
```

Comment:
```suggestion
            else if (!s_targetScopesMap.TryGetValue(scopeString, out targetScope))
```

Reply:
These always seem to stay back after refactoring code :(... Waiting for you to grey these out as unused suppressions :-)

Reply:
I'll fix this one.